x-pack/plugins/stack_connectors/public/components/text_area_with_autocomplete.tsx

```diff
@@ -0,0 +1,328 @@
+  const [selectableHasFocus, setSelectableHasFocus] = useState(false);
+  const [searchWord, setSearchWord] = useState<string | null>(null);
+
+  const optionsToShow: EuiSelectableOption[] = useMemo(() => {
+    return matches?.map((variable) => ({
+      label: variable,
+      data: {
+        description: variable,
+      },
+      'data-test-subj': `${variable}-selectableOption`,
+    }));
+  }, [matches]);
+
+  const onOptionPick = useCallback(
+    (newOptions: EuiSelectableOption[]) => {
+      if (!textAreaRef.current) return;
+      const { value, selectionStart, scrollTop } = textAreaRef.current;
+      const lastSpaceIndex = value.slice(0, selectionStart).lastIndexOf(' ');
+      const lastOpenDoubleCurlyBracketsIndex = value.slice(0, selectionStart).lastIndexOf('{{');
+      const currentWordStartIndex = Math.max(lastSpaceIndex, lastOpenDoubleCurlyBracketsIndex);
+
+      const checkedElement = newOptions.find(({ checked }) => checked === 'on');
+      if (checkedElement) {
+        const newInputText =
+          value.slice(0, currentWordStartIndex) +
+          '{{' +
+          checkedElement.label +
+          '}}' +
+          value.slice(selectionStart);
+
+        editAction(paramsProperty, newInputText.trim(), index);
+        setMatches([]);
+        textAreaRef.current.focus();
+        // We use setTimeout here, because editAction is async function and we need to wait before it executes
+        setTimeout(() => {
+          if (textAreaRef.current) {
+            textAreaRef.current.selectionStart =
+              currentWordStartIndex + checkedElement.label.length + 4;
+            textAreaRef.current.selectionEnd = textAreaRef.current.selectionStart;
+            textAreaRef.current.scrollTop = scrollTop;
+          }
+        }, 0);
+      }
+    },
+    [editAction, index, paramsProperty]
+  );
+
+  const recalcMenuPosition = useCallback(() => {
+    if (!textAreaRef.current) return;
+    const newPosition = getCaretCoordinates(
+      textAreaRef.current,
+      textAreaRef.current.selectionStart
+    );
+    const textAreaClientRect = textAreaRef.current?.getBoundingClientRect();
+
+    const top =
+      textAreaClientRect.top - textAreaRef.current.scrollTop + newPosition.top + newPosition.height;
+    const left = textAreaClientRect.left + window.pageXOffset;
+    const height = newPosition.height;
+    const width = textAreaClientRect.width;
+    setPopupPosition((old) =>
+      old.top !== top || old.left !== left || old.width !== width || old.height !== height
+        ? { top, left, width, height }
+        : old
+    );
+  }, [setPopupPosition, textAreaRef]);
+
+  useEffect(() => {
+    if (!isListOpen) return;
+    // E.g. we have a case, that warning appeared and disappeared from time to time, but without recalculation of
+    // a menu position, a menu will be in a wrong place. To prevent such cases I recalculate a menu position every 300 ms.
+    const interval = setInterval(recalcMenuPosition, 300);
```

Comment:
There also seems to be a small bug where we're not limiting the menu to the bounds of the text area, so if we scroll out of it we see the menu still: 


https://github.com/elastic/kibana/assets/74562234/884af170-5531-4f79-b04f-3c74d689dec9

Reply:
We do not doing this because the menu stays in one place. All our menus behaves this way.

The problem was that we have warning (yellow banner), which can appeared and disappeared, during the writing in textarea.  So rule form rendered already, but then warning disappeared during writing and we need to trigger somehow recalculation of autocomplete menu position.
I've raised the problem with E&C about banner and they have a ticket. But I thought that even when it'll be fixed, new similar thing can appeared, so it safer to have a way to look if autocomplete menu needs to be moved somehow.

Reply:
Thanks for finding a bug. Fixed it here: https://github.com/elastic/kibana/pull/158454/commits/d06894ad163d356bbbbd187b9de156abf562e48b

Reply:
> The problem was that we have warning (yellow banner), which can appeared and disappeared, during the writing in textarea. So rule form rendered already, but then warning disappeared during writing and we need to trigger somehow recalculation of autocomplete menu position.

Can we do this in response to the yellow banner triggering? Perhaps `useEffect`?

Reply:
So I played with the component a little bit, realized that it is true that the warning banner is not helping by always showing and hidding the warning message when the mustache template is not valid most likely because user just type `{{` and they the mustache template become invalid. I think if the mustache template is invalid, we should just show the warning anyway to avoid the Christmas effect here.

Also I realized that the calculation to determine to show or not show the autocomplete options was a little bit off and also why we always need to do the recalculation twice. I did some work to avoid the `setInterval` and other calculation to be more precise. @guskovaue and @JiaweiWu  please let me know what do you think? here the gist of the change https://gist.github.com/XavierM/e30a087573ee0fcb5bfa46ec88556477

Reply:
Also, it be nice to add some validation to make sure that the mustache template is valid to help our user. In another pR of course ;)